apps/wallet/src/engine/runners/solana/history/index.ts

```diff
@@ -0,0 +1,428 @@
+import { getTokenMetadata, solMetadata } from '../metadata';
+import type {
+	AccountBalanceMap,
+	ParsedTokenAccountWithAddress,
+	TransactionBalances,
+} from '../types';
+
+import { extractSwapBalance, JUPITER_V6_PROGRAM_ID } from './swapHistory';
+
+const historyLimit = 20;
+
+export const getTransactionsHistory = async (
+	connection: Connection,
+	wallet: PublicKey,
+	accounts: ParsedTokenAccountWithAddress[],
+) => {
+	const tokenAccounts = accounts
+		.filter((account) => account.tokenAmount.decimals > 0)
+		.map((account) => account.publicKey);
+
+	tokenAccounts.unshift(wallet);
+
+	const signatures = await getTransactionSignatures(connection, tokenAccounts);
+	signatures.splice(historyLimit, signatures.length - historyLimit);
+
+	const txPromises = signatures.map(async (s) => {
+		return throttle(async () => {
+			// don't use getParsedTransactions because
+			// it uses batch rpc request failed by rpc limit without retrying
+			// we need to use getParsedTransaction for separately retry
+			const tx = await connection.getParsedTransaction(s, {
+				maxSupportedTransactionVersion: 0,
+				commitment: 'finalized',
+			});
+
+			const txDoc = await constructTransactionHistoryDocument(
+				connection,
+				tx as ParsedTransactionWithMeta,
+				wallet,
+			);
+
+			if (!txDoc) return;
+			await storage.upsert<TransactionHistoryDocument>(txDoc._id, async () => {
+				return txDoc;
+			});
+		})();
+	});
+
+	await Promise.all(txPromises);
+
+	removeOldHistories();
+};
+
+const getTransactionSignatures = async (
+	connection: Connection,
+	accounts: PublicKey[],
+) => {
+	const signatures = (
+		await Promise.all(
+			accounts.map(async (account) => {
+				const sigs = await throttle(() => {
+					return connection.getSignaturesForAddress(account, {
+						limit: historyLimit,
+					});
+				})();
+
+				return sigs;
+			}),
+		)
+	)
+		.flat()
+		.sort((sig1, sig2) =>
+			(sig1?.blockTime || 0) < (sig2?.blockTime || 0) ? 1 : -1,
+		)
+		.map((s) => s.signature);
+
+	return [...new Set(signatures)];
+};
+
+const constructTransactionHistoryDocument = async (
+	connection: Connection,
+	parsedTransaction: ParsedTransactionWithMeta,
+	wallet: PublicKey,
+): Promise<
+	TransactionHistoryDocument<
+		SolanaTransferHistoryV2 | SolanaSwapHistoryV2 | SolanaUnknownHistory
+	>
+> => {
+	const { meta, transaction, blockTime } = parsedTransaction;
+	const walletAddress = wallet.toString();
+
+	const _id = `${walletAddress}/history/${transaction.signatures[0]}`;
+	const signature = transaction.signatures[0];
+	const id = signature;
+	const date = blockTime ? new Date(blockTime * 1000) : new Date();
+	const status = !meta?.err ? 'Success' : 'Failed';
+	const tokenForFee = getTransactionTokenFee();
+	const fee = (meta?.fee || 0) / 10 ** 9;
+	const balances = await getTransactionBalances(
+		connection,
+		parsedTransaction,
+		walletAddress,
+	);
+
+	if (!tokenForFee || balances.transactionType === 'Unknown') {
+		return {
+			_id,
+			id,
```

Comment:
why do we need two id fields

Reply:
just a legacy shape of history document, I removed it